pallets/settlement/src/lib.rs

```diff
@@ -1609,101 +1644,124 @@ impl<T: Config> Module<T> {
         Ok(details)
     }
 
-    /// Execute the instruction with `instruction_id`, pruning it on success.
-    /// On error, set the instruction status to failed.
-    fn execute_instruction_retryable(id: InstructionId) -> Result<u32, DispatchError> {
-        let result = Self::execute_instruction(id);
-        if result.is_ok() {
-            Self::prune_instruction(id, true);
-        } else if <InstructionDetails<T>>::contains_key(id) {
-            InstructionStatuses::<T>::insert(id, InstructionStatus::Failed);
-        }
-        result
+    /// Executes the instruction of the given `id`. If the execution succeeds, the instruction gets pruned,
+    /// otherwise the instruction status is set to failed.
+    fn execute_instruction_retryable(
+        id: InstructionId,
+        weight_meter: &mut WeightMeter,
+    ) -> DispatchResult {
+        if let Err(e) = Self::execute_instruction(id, weight_meter) {
+            if <InstructionDetails<T>>::contains_key(id) {
+                InstructionStatuses::<T>::insert(id, InstructionStatus::Failed);
+            }
```

Comment:
```suggestion
            InstructionStatuses::<T>::insert(id, InstructionStatus::Failed);
```
I think the `contains_key` check is from before the status field was moved out of the details.